session/utils.go

```diff
@@ -0,0 +1,213 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package session
+
+import (
+	"encoding/json"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/tidb/util/hack"
+)
+
+func reserveBuffer(buf []byte, appendSize int) []byte {
+	newSize := len(buf) + appendSize
+	if cap(buf) < newSize {
+		newBuf := make([]byte, len(buf)*2+appendSize)
+		copy(newBuf, buf)
+		buf = newBuf
+	}
+	return buf[:newSize]
+}
+
+// escapeBytesBackslash will escape []byte into the buffer, with backslash.
+func escapeBytesBackslash(buf []byte, v []byte) []byte {
+	pos := len(buf)
+	buf = reserveBuffer(buf, len(v)*2)
+
+	for _, c := range v {
+		switch c {
+		case '\x00':
+			buf[pos] = '\\'
+			buf[pos+1] = '0'
+			pos += 2
+		case '\n':
+			buf[pos] = '\\'
+			buf[pos+1] = 'n'
+			pos += 2
+		case '\r':
+			buf[pos] = '\\'
+			buf[pos+1] = 'r'
+			pos += 2
+		case '\x1a':
+			buf[pos] = '\\'
+			buf[pos+1] = 'Z'
+			pos += 2
+		case '\'':
+			buf[pos] = '\\'
+			buf[pos+1] = '\''
+			pos += 2
+		case '"':
+			buf[pos] = '\\'
+			buf[pos+1] = '"'
+			pos += 2
+		case '\\':
+			buf[pos] = '\\'
+			buf[pos+1] = '\\'
+			pos += 2
+		default:
+			buf[pos] = c
+			pos++
+		}
+	}
+
+	return buf[:pos]
+}
+
+// escapeStringBackslash will escape string into the buffer, with backslash.
+func escapeStringBackslash(buf []byte, v string) []byte {
+	return escapeBytesBackslash(buf, hack.Slice(v))
+}
+
+// EscapeSQL will escape input arguments into the sql string, doing necessary processing.
+// It works like printf() in c, there are following format specifiers:
+// 1. %?: automatic conversion by the type of arguments. E.g. []string -> ('s1','s2'..)
+// 2. %%: output %
+// 3. %n: for identifiers, for example ("use %n", db)
+// But it does not prevent you from doing EscapeSQL("select '%?", ";SQL injection!;") => "select '';SQL injection!;'".
+// It is still your responsibility to write safe SQL.
+func EscapeSQL(sql string, args ...interface{}) (string, error) {
+	buf := make([]byte, 0, len(sql))
+	argPos := 0
+	for i := 0; i < len(sql); i++ {
+		q := strings.IndexByte(sql[i:], '%')
+		if q == -1 {
+			buf = append(buf, sql[i:]...)
+			break
+		}
+		buf = append(buf, sql[i:i+q]...)
+		i += q
+
+		ch := byte(0)
+		if i+1 < len(sql) {
+			ch = sql[i+1] // get the specifier
+		}
```

Comment:
if the last char of the string is a '%', it's truncated?

Reply:
No. It will not do anything if it it is not one of the known specifiers. I can add a test.